test/e2e/network/endpointslice.go

```diff
@@ -299,6 +303,237 @@ var _ = SIGDescribe("EndpointSlice", func() {
 		deleteEndpointSlices(cs, f.Namespace.Name, svc2)
 		expectEndpointsAndSlices(cs, f.Namespace.Name, svc2, []*v1.Pod{pod1, pod2}, 2, 2, true)
 	})
+
+	// Ensure that garbage collection works well when a Service is deleted and
+	// its EndpointSlices are being rapidly updated.
+	// For more context: https://github.com/kubernetes/kubernetes/issues/91287
+	ginkgo.It("should not get in the way of Service owner garbage collection", func() {
+		sharedLabels := map[string]string{"test-garbage-collection": "for-service"}
+
+		deployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(context.TODO(), &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "test-garbage-collection",
+				Labels: sharedLabels,
+			},
+			Spec: appsv1.DeploymentSpec{
+				Selector: &metav1.LabelSelector{MatchLabels: sharedLabels},
+				Replicas: utilpointer.Int32Ptr(1),
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: sharedLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "container1",
+								Image: imageutils.GetE2EImage(imageutils.Nginx),
+								Ports: []v1.ContainerPort{{
+									Name:          "example-name",
+									ContainerPort: int32(3000),
+								}},
+							},
+						},
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+
+		ownerRef := metav1.NewControllerRef(deployment, schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"})
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "example-int-port",
+				Labels:          sharedLabels,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		esWatch, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).Watch(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+		framework.ExpectNoError(err, "Error watching EndpointSlices")
+		defer esWatch.Stop()
+
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		err = cs.AppsV1().Deployments(f.Namespace.Name).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		})
+		framework.ExpectNoError(err, "Error deleting %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		// There's some delay between when a Service is deleted and the
+		// EndpointSlice informer is updated with that information. This
+		// allows for up to 1 second of delay.
```

Comment:
this seems likely to be flaky... is this trying to compensate for the fact that GC can observe and delete the endpointslice, and the endpointslicecontroller can observe that deletion and react to it before it observes the service deletion?

Reply:
would it be simpler to place an upper bound on the number of recreations (maybe 2?) and amount of time we'll accept (maybe 5 seconds) before we expect endpointslices to stop being recreated, and a time limit (maybe a minute?) for the service and all endpointslices to be gone?

Reply:
That makes sense. I've updated this test to use that kind of approach.